common/argparse.h

```diff
@@ -0,0 +1,893 @@
+//
+// ## Terminology used by this library
+//
+// _Argument_ or _arg_: One of the parsed components of the command line.
+//
+// _Option_: A _named_ argument starting with `--` to configure some aspect of
+// the tool. These often have both long spellings that start with `--` and
+// short, single character spellings that start with `-` and can be bundled with
+// other single character options.
+//
+// _Flag_: A boolean or binary option. These can only be enabled or disabled.
+//
+// _Positional argument_: An argument that is not named but is identified based
+// on the order in which it is encountered in the command line, with options
+// removed. Only a leaf command can contain positional arguments.
+//
+// _Command_: The container of options, subcommands, and positional arguments to
+// parse and an action to take when successful.
+//
+// _Leaf command_: A command that doesn't contain subcommands. This is the only
+// kind of command that can contain positional arguments.
+//
+// _Subcommand_: A command nested within another command and identified by a
+// specific name that ends the parsing of arguments based on the parent and
+// switches to parse based on the specific subcommand's options and positional
+// arguments. A command with subcommands cannot parse positional arguments.
+//
+// _Action_: An open-ended callback, typically reflecting a specific subcommand
+// being parsed. Can either directly perform the operation or simply mark which
+// operation was selected.
+//
+// _Meta action_: An action fully handled by argument parsing such as displaying
+// help, version information, or completion.
+//
+// Example command to illustrate the different components:
+//
+//     git --no-pager clone --shared --filter=blob:none my-repo my-directory
+//
+// This command breaks down as:
+// - `git`: The top-level command.
+// - `--no-pager`: A negated flag on the top-level command (`git`).
+// - `clone`: A subcommand.
+// - `--shared`: A positive flag for the subcommand (`clone`).
+// - `--filter=blob:none`: An option named `filter` with a value `blob:none`.
+// - `my-repo`: The first positional argument to the subcommand.
+// - `my-directory`: the second positional argument to the subcommand.
+//
+// **Note:** while the example uses a `git` command to be relatively familiar
+// and well documented, this library does not support the same flag syntaxes as
+// `git` or use anything similar for its subcommand dispatch. This example is
+// just to help clarify the terminology used, and carefully chosen to only use a
+// syntax that overlaps with this library's parsed syntax.
+//
+// ## Options and flags
+//
+// The option syntax and behavior supported by this library is designed to be
+// strict and relatively simple while still supporting a wide range of expected
+// use cases:
+//
+// - All options must have a unique long name that is accessed with a `--`
+//   prefix. The name must consist of characters in the set [-a-zA-Z0-1], and it
```

Comment:
```suggestion
//   prefix. The name must consist of characters in the set [-a-zA-Z0-9], and it
```
I assume this is a typo?